src/hotspot/share/prims/jvmtiEnvBase.cpp

```diff
@@ -1944,6 +1940,55 @@ MultipleStackTracesCollector::allocate_and_fill_stacks(jint thread_count) {
          "the last copied frame info must be the last record");
 }
 
+// Supports platform and virtual threads.
+// JvmtiVTMSTransitionDisabler is always set by this function.
+void
+JvmtiHandshake::execute(JvmtiUnitedHandshakeClosure* hs_cl, jthread target) {
+  JavaThread* current = JavaThread::current();
+  HandleMark hm(current);
+
+  JvmtiVTMSTransitionDisabler disabler(target);
+  ThreadsListHandle tlh(current);
+  JavaThread* java_thread = nullptr;
+  oop thread_obj = nullptr;
+
+  jvmtiError err = JvmtiEnvBase::get_threadOop_and_JavaThread(tlh.list(), target, &java_thread, &thread_obj);
+  if (err != JVMTI_ERROR_NONE) {
+    hs_cl->set_result(err);
+    return;
+  }
+  Handle target_h(current, thread_obj);
+  execute(hs_cl, &tlh, java_thread, target_h);
+}
+
+// Supports platform and virtual threads.
+// A virtual thread is always identified by the target_h oop handle.
+// The target_jt is always nullptr for an unmounted virtual thread.
+// JvmtiVTMSTransitionDisabler has to be set before call to this function.
+void
+JvmtiHandshake::execute(JvmtiUnitedHandshakeClosure* hs_cl, ThreadsListHandle* tlh,
+                        JavaThread* target_jt, Handle target_h) {
+  bool self = target_jt == JavaThread::current();
+
+  hs_cl->set_self(self);           // needed when suspend is required for non-current target thread
+  hs_cl->set_target_h(target_h);   // need this to differentiate between virtual and carrier thread
+
+  if (java_lang_VirtualThread::is_instance(target_h())) { // virtual thread
+    if (!JvmtiEnvBase::is_vthread_alive(target_h())) {
+      return;
+    }
+    if (target_jt == nullptr) {    // unmounted virtual thread
+      hs_cl->do_vthread(target_h); // execute handshake closure callback on current thread directly
+    }
+  }
+  if (target_jt != nullptr) {      // mounted virtual or platform thread
+    if (self) {                    // target thread is current
+      hs_cl->do_thread(target_jt); // execute handshake closure callback on current thread directly
+    } else {
+      Handshake::execute(hs_cl, tlh, target_jt); // delegate to Handshake implementation
+    }
```

Comment:
Every implementation of JvmtiUnitedHandshakeClosure has to check if the target thread is virtual and call do_vthread manually.
I'd suggest to handle this by proxy class, something like
```suggestion
    class Adapter : public HandshakeClosure {
      JvmtiUnitedHandshakeClosure* _hs_cl;
      Handle _target_h;
    public:
      Adapter(JvmtiUnitedHandshakeClosure* hs_cl, Handle target_h)
          : HandshakeClosure(hs_cl->name()), _hs_cl(hs_cl), _target_h(target_h) {}
      virtual void do_thread(Thread* thread) {
        if (java_lang_VirtualThread::is_instance(_target_h())) { // virtual thread
          _hs_cl->do_vthread(_target_h);
        } else {
          _hs_cl->do_thread(target);
        }
      }
    } adapter(hs_cl, target_h);

    if (self) {                    // target thread is current
      adapter.do_thread(target_jt); // execute handshake closure callback on current thread directly
    } else {
      Handshake::execute(&adapter, tlh, target_jt); // delegate to Handshake implementation
    }
```

Reply:
Thank you for the suggestion! Agreed, this should help to get rid of this duplication/ugliness.

Reply:
Added the fix suggested by Alex.